### PR TITLE
(DOCSP-30575): POST /conversations/ route

### PR DESCRIPTION
**Pull request by mongodben**
Add `POST /conversations/` route. refactored the conversations router code for separation of concern and testability.

### CODE REVIEW

File: chat-server/src/routes/conversations/index.ts [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+}: ConversationsRouterParams<OpenAiStreamingResponse, OpenAiAwaitedResponse>) {
+  const conversationsRouter = Router();
+
+  const createConversationRoute = makeCreateConversationRoute({
+    conversations,
+  });
+  /**
+   * Create new conversation.
+   */
+  conversationsRouter.post("/", createConversationRoute);
```
Comment (nlarew) [inline chat-server/src/routes/conversations/index.ts:39]:
```suggestion
  /**
   * Create new conversation.
   */
  conversationsRouter.post("/", makeCreateConversationRoute({ conversations }));
```

Can we inline the call here like we do on the other endpoints?

File: chat-server/src/routes/conversations/rateMessage.ts [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    conversationId: string;
+    messageId: string;
+  };
+  body: {
+    conversation: Message[];
+    id: string;
+    ip_address: string;
+    message_index: number;
+    rating: boolean;
+  };
```
Comment (nlarew) [inline chat-server/src/routes/conversations/rateMessage.ts:23]:
```suggestion
  body: {
    rating: boolean;
  };
```

We only need `rating` here

File: chat-server/src/routes/conversations/utils.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { Conversation } from "../../services/conversations";
+
+export interface MessageResponse {
+  id: string;
+  role: string;
+  content: string;
+  rating?: boolean;
+  timeCreated: number;
```
Comment (nlarew) [inline chat-server/src/routes/conversations/utils.ts:8]:
```suggestion
  createdAt: number;
```

We're using `createdAt` (and `updatedAt` etc) in the spec

File: chat-server/src/routes/conversations/utils.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    id: conversation._id.toString(),
+    messages: conversation.messages.map((message) => ({
+      id: message.id.toString(),
+      role: message.role,
+      content: message.content,
+      rating: message.rating,
+      timeCreated: message.timeCreated.getTime(),
+    })),
+    timeCreated: conversation.timeCreated.getTime(),
+  };
```
Comment (nlarew) [inline chat-server/src/routes/conversations/utils.ts:29]:
```suggestion
  return {
    id: conversation._id.toString(),
    messages: conversation.messages.map((message) => ({
      id: message.id.toString(),
      role: message.role,
      content: message.content,
      rating: message.rating,
      createdAt: message.createdAt.getTime(),
    })),
    createdAt: conversation.createdAt.getTime(),
  };
```

We're using `createdAt` (and `updatedAt` etc) in the spec

File: chat-server/src/services/dataStreamer.ts [hunk 5]
```diff
@@ -12,12 +14,15 @@ interface AnswerParams {
...
 }
 
-class DataStreamer {
+export interface DataStreamerServiceInterface {
+  answer(params: AnswerParams): Promise<string>;
+}
+export class DataStreamerService {
   // NOTE: for example streaming data, see https://github.com/openai/openai-node/issues/18#issuecomment-1369996933
   async answer({ res, answer, conversation, chunks }: AnswerParams) {
     // TODO: do stuff with the response
```
Comment (nlarew) [inline chat-server/src/services/dataStreamer.ts:23]: Streaming is out of scope for this PR right?

File: chat-server/src/routes/conversations/index.ts (hunk 1 above)
Comment (mongodben) [reply chat-server/src/routes/conversations/index.ts:39]: yeah, that was a legacy of some debugging. good catch

File: design_docs/openapi.yml [hunk 6]
```diff
@@ -146,50 +146,44 @@ components:
...
             $ref: "#/components/schemas/Message"
-        user_ip:
-          type: string
-          format: ipv4
-          description: The IP address of the user performing the conversation.
-        createdAt:
-          type: integer
-          description: The date the conversation was created.
-        updatedAt:
+        timeCreated:
```
Comment (nlarew) [inline design_docs/openapi.yml:177]: Why change the spec schema?

File: chat-server/src/services/dataStreamer.ts (hunk 5 above)
Comment (mongodben) [reply chat-server/src/services/dataStreamer.ts:23]: yeah, this was just for tsc

File: design_docs/openapi.yml (hunk 6 above)
Comment (mongodben) [reply design_docs/openapi.yml:177]: b/c right now we're storing `timeCreated` in DB so i thought it'd be easier to keep API and DB consistent on this.

Comment (nlarew) [reply design_docs/openapi.yml:177]: Why are we using `timeCreated` in the db? The spec schemas for those also use `createdAt` and `updatedAt`.

Comment (nlarew) [reply design_docs/openapi.yml:177]: Ultimately it doesn't matter that much but we should try to stick to the specs IMO

Comment (mongodben) [reply design_docs/openapi.yml:177]: if there was a reason why, i don't remember. will update everything to `(created|updated)At`

File: chat-server/src/routes/conversations/utils.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  id: string;
+  messages: MessageResponse[];
+  createdAt: number;
+}
+export function convertConversationToResponse(
+  conversation: Conversation
+): ConversationResponse {
+  conversation.messages.shift(); // Remove the system prompt
+  return {
+    id: conversation._id.toString(),
```
Comment (nlarew) [inline chat-server/src/routes/conversations/utils.ts:20]:
```suggestion
export interface ConversationResponse {
  _id: string;
  messages: MessageResponse[];
  createdAt: number;
}
export function convertConversationToResponse(
  conversation: Conversation
): ConversationResponse {
  conversation.messages.shift(); // Remove the system prompt
  return {
    _id: conversation._id.toString(),
```

Spec has `_id` for the root-level document ID.

**Review by nlarew — APPROVED** [review]
LGTM after one more small suggested change

File: chat-server/src/routes/conversations/utils.ts (hunk 7 above)
Comment (mongodben) [reply chat-server/src/routes/conversations/utils.ts]: ok, changing spec back as well

**Pull request merged**